packages/block-editor/src/components/block-patterns-list/index.js

```diff
@@ -18,6 +18,9 @@ import {
 import { useInstanceId } from '@wordpress/compose';
 import { __ } from '@wordpress/i18n';
 import { Icon, symbol } from '@wordpress/icons';
+// ignore restricted import ESLint error for apiFetch
+// eslint-disable-next-line no-restricted-imports
+import apiFetch from '@wordpress/api-fetch';
```

Comment:
The block editor can't depend on apiFetch as it's a generic JS package. What we should do here is to provide a block editor setting that is called `renderPreview` or something that is basically a component that takes a list of blocks to render their preview. And if this setting is not provided, we should keep the existing preview behavior (third-party editors...)

Reply:
Yes I've commented as well that this is a known issue. The eslint disable is for demo purposes. 

If the endpoint gets a list of patterns and returns a list of rendered html then the patterns that are passed to this component list can include their rendered html and this fetch would go away. Why would we need an editor setting?

Reply:
This component is about rendering previews regardless of whether they come from pattern, the example of a block type or any random block list. Adding the rendered html to patterns won't solve all the use-cases.

Reply:
The block preview component optionally gets the HTML. It is being passed this HTML depending on the context. In this case only it comes from block patterns list which gets it from whomever wants to render a pattern list. So yes, what I am saying is I only want to solve the pattern preview case but the low level block preview will be ready for future situations. I don't understand what the editor setting would be for, it's probably obvious :D

Reply:
Why would a BlockPreview component get HTML, it's not block preview anymore, it's just an iframe in that case and there's no need to use this component.

For me, we should solve all the use-cases at once because they're just as problematic in terms of performance.

>  I don't understand what the editor setting would be for, it's probably obvious :D

This is some pseudo code, to explain the idea.

```js
const BlockPreview = ( blocks ) => {
   const html = useSelect( select => select( 'core/data' ).getHtmlForBlocks( blocks ), [ blocks ] );
   return <Iframe>{html ?? <Spinner />}</Iframe>
}

<BlockEditorProvider settings={ ...settings, BlockPreview } />
```

and the `BlockPreview` component could do something like:

```js
const { BlockPreview } = useSelect( select => select( 'core/block-editor' ).getSettings().BlockPreview, [] );

return BlockPreview ? <BlockPreview blocks={blocks} /> : <CurrentBlockPreviewCode blocks={blocks} />
```

Reply:
> Why would a BlockPreview component get HTML, it's not block preview anymore,

Interesting, I think the function of the component is to preview blocks, wether it previews them via some html in an iframe or via a rendered block editor is irrelevant to me.

However I don't see any issue with having two distinct components for blocks and html. I wonder if there will be cases when we'll need to render a block editor and server side rendering will not be enough ..

Either way I'll go the separate component route. But for now I'll focus on the endpoint.

Reply:
> So yes, what I am saying is I only want to solve the pattern preview case but the low level block preview will be ready for future situations.

I wonder why we shouldn't just alter the patterns APIs to make them return rendered `html` too? Then in `<BlockPreview>` (or a different `<StaticBlockPreview>` if needed), we short-circuit the original block editor path if `html` is passed as a prop and just render the content via `dangerouslySetInnerHTML` instead.

Reply:
I'm not against two components to render the previous, but in that case, It's not even a `StaticBlockPreview` it's just `Iframe`

Reply:
I think there's still some hot-wiring to make an `<Iframe>` `<StaticBlockPreview>`. The injected styles and scaled size as mentioned by @draganescu for instance, but also like `pointer-events: none` and other stuff that only makes sense in the "preview" context. We probably will end up abstracting them into a different component, but the name should be more explicit than `<Iframe>`.

Reply:
Yes, as @kevin940726 describes the helper components that block preview needs will have to be abstracted and provided by a 3rd package if we want to give editor implementations the ability to implement their own previewer - like this PR demonstrates for the site editor.

I don't think we should modify the patterns API - previewing is very special - again as this PR demonstrates - in the sense that we need to build some context for the rendering to be as close to a real preview as possible - which complicates the process enough to warrant a special endpoint. I also agree with @youknowriad that if should have a blanket solution for all previews: patterns, blocks, block styles.

Reply:
> previewing is very special - again as this PR demonstrates - in the sense that we need to build some context for the rendering to be as close to a real preview as possible - which complicates the process enough to warrant a special endpoint.

Could you elaborate what specifically is the special handling we need to do in a separate API request, as opposed to maybe a PHP function call instead? There's nothing stopping us from providing both the endpoint and patching the patterns API too. I guess the main concern I have here is that a round-trip time for each preview depends on network condition and might just be slower for some users (high-end device but with slow internet). Not to mention that we're not doing batching and we'll soon reach the cap of maximum concurrent requests the browser can make at a time. We're also not doing error management and retrying which make this a little bit more complicated. I think it'll be easier if we can just move them all to the patterns API so that it can also be cached correctly too (any reason we're using `POST` for this endpoint that can't be cached by the browser?).

Reply:
I think your observations are correct. There's nothing stopping us from providing both the endpoint and patching the patterns API too, indeed. 

> what specifically is the special handling we need to do in a separate API request, as opposed to maybe a PHP function call instead

It's not specific handling it's specific logic that we require to execute to create a server context for rendering properly. We could have shared procedures (a function) between patterns and other custom endpoints, sure.

> any reason we're using POST for this endpoint 

We send this big blob of blocks. If this were part of the GET patterns this would also not be required.